guides/api-style.rst

```diff
@@ -0,0 +1,125 @@
+
+Note that it is not a guide to *code* style, only API design.
+
+The Hypothesis style evolves over time, and earlier strategies in particular
+may not be consistent with this style, and we've tried some experiments
+that didn't work out, so this style guide is more normative than descriptive
+and existing APIs may not match it. Where relevant, backwards compatibility is
+much more important than conformance to the style.
+
+~~~~~~~~~~~~~~~~~~
+General Guidelines
+~~~~~~~~~~~~~~~~~~
+
+* When writing extras modules, consistency with Hypothesis trumps consistency
+  with the library you're integrating with.
+* *Absolutely no subclassing as part of the public API*
+* We should not strive too hard to be pythonic, but if an API seems weird to a
+  normal Python user we should see if we can come up with an API we like as
+  much but is less weird.
+* Code which adds a dependency on a third party package should be put in a
+  hypothesis.extra module.
+* Complexity should not be pushed onto the user. An easy to use API is more
+  important than a simple implementation.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~
+Guidelines for strategies
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* A strategy function should be somewhere between a recipe for how to build a
+  value and a range of valid values.
+* It should not include distribution hints. The arguments should only specify
+  how to produce a valid value, not statistical properties of values.
+* Strategies should try to paper over non-uniformity in the underlying types
+  as much as possible (e.g. ``hypothesis.extra.numpy`` has a number of
+  workarounds for numpy's odd behaviour around object arrays).
+
+~~~~~~~~~~~~~~~~~
+Argument handling
+~~~~~~~~~~~~~~~~~
+
+We have a reasonably distinctive style when it comes to handling arguments:
+
+* Arguments must be validated to the greatest extent possible. Hypothesis
+  should reject bad arguments with an InvalidArgument error, not fail with an
+  internal exception.
+* We make extensive use of default arguments. If an argument could reasonably
+  have a default, it should.
+* Exception to the above: Strategies for collection types should *not* have a
+  default argument for element strategies.
+* Interacting arguments (e.g. arguments that must be in a particular order, or
+  where at most one is valid, or where one argument restricts the valid range
+  of the other) are fine, but when this happens the behaviour of defaults
+  should automatically be adjusted. e.g. if the normal default of an argument
+  would become invalid, the function should still do the right thing if that
+  default is used.
+* Where the actual default used depends on other arguments, the default parameter
+  should be None.
+* It's worth thinking about the order of arguments: the first one or two
+  arguments are likely to be passed positionally, so try to put values there
+  where this is useful and not too confusing.
+* Arguments should not be "a value or a strategy for generating that value".
```

Comment:
I would reverse these two points: first suggest making an argument a strategy, then warn away from value-or-strategy.  I think it just flows better that way.